File: dotnet/Allmark/Render/TextRenderer.cs
namespace Allmark.Render;

using Allmark.Types;

public static class TextRenderer
{
    public static OutputRenderer Create()
    {
        return new OutputRenderer
        {
            Name = "text",
            Render = Render,
        };
    }

    public static void Render(MarkdownNode node, RendererState state, bool? decode = true)
    {
        var content = node.Content;
        var scanDecode = decode == true;

        // Fast path: if none of the special characters are present, output as-is
        var needsProcessing = scanDecode
            ? content.AsSpan().IndexOfAny("&<>\"\\") >= 0
            : content.AsSpan().IndexOfAny("&<>\"") >= 0;
        if (!needsProcessing)
        {
            state.Output.Append(content);
            return;
        }

        if (scanDecode)
        {
            content = Utils.DecodeEntities(content);
            content = Utils.EscapePunctuation(content);
        }
        content = Utils.EscapeHtml(content);

        state.Output.Append(content);
    }
}

File: web/src/render-console/textRenderer.ts
import type ConsoleRendererState from "../types/ConsoleRendererState";
import type MarkdownNode from "../types/MarkdownNode";
import type Renderer from "../types/Renderer";
import type RendererState from "../types/RendererState";

const renderer: Renderer = {
	name: "text",
	render,
};
export default renderer;

function render(node: MarkdownNode, state: RendererState): void {
	const s = state as ConsoleRendererState;
	s.output += node.content;
}
